db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/HashableJoinStrategy.java

```diff
@@ -135,8 +137,26 @@ public boolean feasible(Optimizable innerTable,
             cd = innerTable.getCurrentAccessPath().getConglomerateDescriptor();
         }
 
-        /* Look for equijoins in the predicate list */
-        hashKeyColumns = findHashKeyColumns(innerTable, cd, predList);
+        /* Exclude predicates that are not useful in probing hash table built from right table.
+         * Make sure the same logic applies here and in divideUpPredicateLists(). Otherwise,
+         * this routine could return a hash-based join strategy is feasible but later in
+         * modifying access paths, no hash columns is found.
+         */
+        if (predList != null) {
+            JBitSet joinedTablesInCurrentQueryBlock = optimizer.getAssignedTableMap();
+            PredicateList predListCopy = new PredicateList();
+            PredicateList nljPreds = new PredicateList();
+            ContextManager cm = ((QueryTreeNode) innerTable).getContextManager();
+            predListCopy.setContextManager(cm);
+            nljPreds.setContextManager(cm);
+            predList.copyPredicatesToOtherList(predListCopy);
+            predListCopy.transferPredicates(nljPreds, innerTable.getReferencedTableMap(), innerTable, joinedTablesInCurrentQueryBlock);
```

Comment:
It may cause extra memory and processing overhead to build new predicate lists every time the optimizer considers a new hashable join strategy.  It would be better to just pass joinedTableSet into findHashKeyColumns and hasOptimizableEquijoin.  Then we can write a new function that both transferPredicates and hasOptimizableEquijoin can use to test a predicate:

> JBitSet getMaskedReferenceSet(Predicate predicate, JBitSet joinedTableSet) {
>     JBitSet maskedReferencedSet = (JBitSet)predicate.getReferencedSet().clone();
>     maskedReferencedSet.and(joinedTableSet);
>     return maskedReferencedSet
> }

In hasOptimizableEquijoinHelper, add:

>             // skip NLJ predicates
>             JBitSet maskedReferencedSet = getMaskedReferenceSet(predicate, joinedTableSet);
>             if (optTable.getReferencedTableMap().contains(maskedReferencedSet)) {
>                 continue;
>             }

The test: `if (optTable.getReferencedTableMap().contains(maskedReferencedSet))`
could also be refactored into a function to be shared between hasOptimizableEquijoinHelper and transferPredicates so that we won't ever modify the logic in one location without modifying the logic in the other location.  If we could find out a way to avoid the call to predicate.getReferencedSet().clone() we could avoid memory to an even greater extent, but I don't know if this is possible.

Reply:
OK, I refactored the code a bit to remove the need of constructing new predicate lists.